### ISSUE TITLE
What is the relationship between steps and image2image strength?

### ISSUE THREAD

**Issue by twalderman** [filipstrand/mflux#86]
I see some dynamic between the 2 but Im not clear on how one affects the other. The steps I define vary in execution and I see a relationship with --init-image-strength.  Insights welcome.

**Comment by anthonywu** [comment]
this is a really good question, and one that I explained in the code review comment here: https://github.com/filipstrand/mflux/pull/77#issuecomment-2408848975

this comes down to terminology: in my implementation of img2img in mflux, the arg name I used was `--init-image-strength`, and the implementation is interpreted as "how strong the initial image influences the outcome", so the "init image strength" arg is named, higher value -> resembles reference image, lower value -> diverges farther from reference image. I think this is intuitive for most people to think about.


```python
    @property
    def init_time_step(self) -> int:
        if self.config.init_image_path is None:
            # text to image, always begin at time step 0
            return 0
        else:
            # we skip to the time step as informed by the init_image_strength
            # the higher the strength number, the more time steps we skip
            strength = max(0.0, min(1.0, self.config.init_image_strength))
            # if the strength is too small to even influence the image
            # help the user round up so the init_image has influence at step 1
            t = max(1, int(self.num_inference_steps * strength))
            return t
```


in StableDiffusion implementation commonly found elsewhere, a common term is called "denoise" strength, and in the [diffusers pipeline_flux_img2img.py implementation](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_img2img.py#L414C1-L423C56) is this:

```py
    def get_timesteps(self, num_inference_steps, strength, device):
        # get the original timestep using init_timestep
        init_timestep = min(num_inference_steps * strength, num_inference_steps)

        t_start = int(max(num_inference_steps - init_timestep, 0))
        timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]
        if hasattr(self.scheduler, "set_begin_index"):
            self.scheduler.set_begin_index(t_start * self.scheduler.order)

        return timesteps, num_inference_steps - t_start
```

where their "denoise **strength**" [arg is described](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_img2img.py#L625
) as

```txt
            strength (`float`, *optional*, defaults to 1.0):
                Indicates extent to transform the reference `image`. Must be between 0 and 1. `image` is used as a
                starting point and more noise is added the higher the `strength`. The number of denoising steps depends
                on the amount of noise initially added. When `strength` is 1, added noise is maximum and the denoising
                process runs for the full number of iterations specified in `num_inference_steps`. A value of 1
                essentially ignores `image`.
```

^ if you name the strength as a "denoise" value, it's interpreted opposite of `--init-image-strength` in `mflux` `0.4.0` release

-----

@filipstrand @Xuzzo I realized that during my implementation I diverged away from @Xuzzo's attempt in https://github.com/filipstrand/mflux/pull/16/files#diff-dc9005328f1b43fe9c39cc56aba83a0927bfef1bd27a78e60707b626312588f9R54 where @Xuzzo pads additional inference steps

```
       self.strength = strength
         if strength <= 0.0 or strength >= 1.0:
             raise ValueError("Strength should be a float between 0 and 1.")

         self.num_total_denoising_steps = int(self.num_inference_steps / (1-strength))
         self.init_timestep = int(self.num_total_denoising_steps - self.num_inference_steps) 
         self.inference_steps = list(range(self.num_total_denoising_steps))[self.init_timestep:]
```

Between the upstream diffusers/stablediffusion implementation and the two attempts here in mflux. I'm not sure which one is best, but maybe we can offer users various ways of specifying the values via mutually exclusive args:

1. `--init-image-strength` - keep implementation as is, let users choose steps, no additional step padding logic
2. `--denoise-strength` - implement like upstream diffusers, but I feel this arg name is unintuitive and "exposes algorithmic details" to the user to consider.

**Comment by twalderman** [comment]
but why do steps vary from the implicitly stated number?  If OI am doing img2img, and I say for instances --step 30 and mflux only uses 8 steps, why is that?

**Comment by anthonywu** [comment]
> but why do steps vary from the implicitly stated number?  If OI am doing img2img, and I say for instances --step 30 and mflux only uses 8 steps, why is that?

The strength fast forwards your generation to the proportion as calculated via (steps * strength). So if you have strength 0.5 and 30 steps you'd only see 15 gen steps, because the first 15 steps (0.5) was filled in with noisy latent of your init image.

**Comment by raysers** [comment]
> this is a really good question, and one that I explained in the code review comment here: [#77 (comment)](https://github.com/filipstrand/mflux/pull/77#issuecomment-2408848975)
> 
> this comes down to terminology: in my implementation of img2img in mflux, the arg name I used was `--init-image-strength`, and the implementation is interpreted as "how strong the initial image influences the outcome", so the way "init image strength" arg is named, higher value -> resembles reference image, lower value -> diverges farther from reference image. I think this is intuitive for most people to think about.
> 
> ```python
>     @property
>     def init_time_step(self) -> int:
>         if self.config.init_image_path is None:
>             # text to image, always begin at time step 0
>             return 0
>         else:
>             # we skip to the time step as informed by the init_image_strength
>             # the higher the strength number, the more time steps we skip
>             strength = max(0.0, min(1.0, self.config.init_image_strength))
>             # if the strength is too small to even influence the image
>             # help the user round up so the init_image has influence at step 1
>             t = max(1, int(self.num_inference_steps * strength))
>             return t
> ```
> 
> in StableDiffusion implementation commonly found elsewhere, a common term is called "denoise" strength, and in the [diffusers pipeline_flux_img2img.py implementation](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_img2img.py#L414C1-L423C56) is this:
> 
> ```python
>     def get_timesteps(self, num_inference_steps, strength, device):
>         # get the original timestep using init_timestep
>         init_timestep = min(num_inference_steps * strength, num_inference_steps)
> 
>         t_start = int(max(num_inference_steps - init_timestep, 0))
>         timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]
>         if hasattr(self.scheduler, "set_begin_index"):
>             self.scheduler.set_begin_index(t_start * self.scheduler.order)
> 
>         return timesteps, num_inference_steps - t_start
> ```
> 
> where their "denoise **strength**" [arg is described](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_img2img.py#L625) as
> 
> ```
>             strength (`float`, *optional*, defaults to 1.0):
>                 Indicates extent to transform the reference `image`. Must be between 0 and 1. `image` is used as a
>                 starting point and more noise is added the higher the `strength`. The number of denoising steps depends
>                 on the amount of noise initially added. When `strength` is 1, added noise is maximum and the denoising
>                 process runs for the full number of iterations specified in `num_inference_steps`. A value of 1
>                 essentially ignores `image`.
> ```
> 
> ^ if you name the strength as a "denoise" value, it's interpreted opposite of `--init-image-strength` in `mflux` `0.4.0` release
> 
> @filipstrand @Xuzzo I realized that during my implementation I diverged away from @Xuzzo's attempt in https://github.com/filipstrand/mflux/pull/16/files#diff-dc9005328f1b43fe9c39cc56aba83a0927bfef1bd27a78e60707b626312588f9R54 where @Xuzzo pads additional inference steps
> 
> ```
>        self.strength = strength
>          if strength <= 0.0 or strength >= 1.0:
>              raise ValueError("Strength should be a float between 0 and 1.")
> 
>          self.num_total_denoising_steps = int(self.num_inference_steps / (1-strength))
>          self.init_timestep = int(self.num_total_denoising_steps - self.num_inference_steps) 
>          self.inference_steps = list(range(self.num_total_denoising_steps))[self.init_timestep:]
> ```
> 
> Between the upstream diffusers/stablediffusion implementation and the two attempts here in mflux. I'm not sure which one is best, but maybe we can offer users various ways of specifying the values via mutually exclusive args:
> 
> 1. `--init-image-strength` - keep implementation as is, let users choose steps, no additional step padding logic
> 2. `--denoise-strength` - implement like upstream diffusers, but I feel this arg name is unintuitive and "exposes algorithmic details" to the user to consider.

I tried the current strength control using the terminal, and it looks very streamlined and efficient due to the reduced steps. There's also another possibility that denoise might better align with the habits of older users. If a dual option could be offered, that would be fantastic

**Comment by anthonywu** [comment]
@raysers that's right, if users want more steps just manually up the step count. For example: 40 steps with strength 0.25 will give you the 30 denoise iterations

**Comment by twalderman** [comment]
My usecase is that I set up a shell script that takes my prompt as a variable and runs it through schnell 2 step and then dev with image strength of .4 and 30 additional steps.  the variables are in the shell script and not passed.  all ofther parameters I prefer are in there as well.   The result is a schnell image refined by Dev giving me the quality that I want with some of the character of schnell dictating the composition Dev uses.  The explanations above have been helpful.  TYVM

**Comment by anthonywu** [comment]
Closing due to current feature status being usable, and the "math" has been explained in this thread.
